helix-term/src/commands.rs

```diff
@@ -2884,12 +2882,37 @@ pub mod insert {
             ..
         }) = &capabilities.completion_provider
         {
-            // TODO: what if trigger is multiple chars long
             if triggers.iter().any(|trigger| trigger.contains(ch)) {
-                cx.editor.clear_idle_timer();
-                super::completion(cx);
+                return true;
             }
         }
+        false
+    }
+
+    fn language_server_completion(cx: &mut Context, ch: char) {
+        use helix_core::chars::char_is_word;
+
+        let config = cx.editor.config();
+        if !config.auto_completion {
+            return;
+        }
+        let (view, doc) = current_ref!(cx.editor);
+        if char_is_word(ch) && doc.savepoint.is_none() {
+            let text = doc.text().slice(..);
+            let cursor = doc.selection(view.id).primary().cursor(text);
+            let mut iter = text.chars_at(cursor);
+            iter.reverse();
+            for _ in 0..config.completion_trigger_len {
+                if iter.next().map_or(true, |c| !char_is_word(c)) {
+                    return;
+                }
+            }
+            cx.editor.reset_idle_timer();
+            return;
+        }
```

Comment:
Completion happens only on the primary cursor? Does the same text gets shown for the rest of the selections?

Reply:
I didn't consider the case. In this case, we should carefully consider what rules are needed